src/support/space.h

```diff
@@ -22,132 +22,65 @@
-// large amounts of overlap between each other. In that case, checking for
-// possible overlaps is logarithmic.
-struct BSPNode {
-  // The minimum width of a span (so that we don't end up emitting a span per
-  // single location or even close to that.
-  const Address MinWidth = 100;
-
+struct DisjointSpans {
+  // A span of form [a, b), i.e., that does not include the end point.
   struct Span {
-    // Left is included, right is not.
     Address left, right;
 
-    bool hasOverlap(const Span& other) {
+    bool checkOverlap(const Span& other) const {
       return !(left >= other.right || right <= other.left);
     }
   };
 
-  // The area we are responsible for.
-  Span area;
-
-  // All the spans that belong to this node.
-  std::vector<Span> mySpans;
-
-  std::unique_ptr<BSPNode> leftChild, rightChild;
-
-  BSPNode(Span area) : area(area) {}
-
-  bool hasOverlap(Span span) {
-    std::vector<BSPNode*> stack;
-    stack.push_back(this);
-    while (!stack.empty()) {
-      auto* curr = stack.back();
-      stack.pop_back();
-      // First compare with existing spans on this node itself, that is, that
-      // overlap both the left and right side.
-      for (auto& existing : curr->mySpans) {
-        if (existing.hasOverlap(span)) {
-          return true;
-        }
-      }
-      // Scan the relevant children.
-      if (curr->hasLeft() && curr->overlapsWithLeft(span)) {
-        stack.push_back(curr->getLeft());
-      }
-      if (curr->hasRight() && curr->overlapsWithRight(span)) {
-        stack.push_back(curr->getRight());
-      }
+  struct SortByLeft {
+    bool operator()(const Span& left, const Span& right) const {
+      return left.left < right.left ||
+             (left.left == right.left && left.right < right.right);
     }
-    return false;
-  }
+  };
 
-  void add(Span span) {
-    BSPNode* curr = this;
-    while (1) {
-      // If this is big enough, then recurse into a smaller child, possibly
-      // creating it.
-      if (curr->getWidth() > MinWidth) {
-        if (curr->entirelyInLeft(span)) {
-          curr = curr->getLeft();
-          continue;
-        }
-        if (curr->entirelyInRight(span)) {
-          curr = curr->getRight();
-          continue;
-        }
+  // The spans seen so far. Guaranteed to be disjoint.
+  std::set<Span, SortByLeft> spans;
+
+  // Adds an item and checks overlap while doing so, returning true if such
+  // overlap exists.
+  bool addAndCheckOverlap(Span span) {
+    // Insert the new span. We can then find its predecessor and successor.
+    // They are disjoint by assumption, so the question is then does the new
+    // span overlap with them, or not.
+    decltype(spans)::iterator iter;
+    bool inserted;
+    std::tie(iter, inserted) = spans.insert(span);
+    if (!inserted) {
+      return true;
+    }
+    if (iter != spans.begin()) {
+      auto before = iter;
+      before--;
+      if (before != spans.end() && before->checkOverlap(span)) {
+        return true;
```

Comment:
Could this be simplified to somthing like this? In particular I don't think you need to check `before != spans.end()`.

```suggestion
    if (iter != spans.begin() && std::prev(iter)->checkOverlao(span)) {
        return true;
```

Reply:
Good point, that is simpler, updated.